server/src/main/java/org/elasticsearch/action/search/SearchQueryRetryThenFetchAsyncAction.java

```diff
@@ -0,0 +1,144 @@
+        // at the end of the search
+        context.addReleasable(resultConsumer);
+
+        boolean hasFetchPhase = request.source() == null ? true : request.source().size() > 0;
+        progressListener.notifyListShards(SearchProgressListener.buildSearchShards(this.shardsIts),
+            SearchProgressListener.buildSearchShards(toSkipShardsIts), context.getClusters(), hasFetchPhase);
+    }
+
+    protected void executePhaseOnShard(final SearchShardIterator shardIt,
+                                       final SearchShardTarget shard,
+                                       final SearchActionListener<SearchPhaseResult> listener) {
+        ShardSearchRequest request = rewriteShardSearchRequest(context.buildShardSearchRequest(shardIt, listener.requestIndex));
+        context.getSearchTransport().sendExecuteQuery(context.getConnection(shard.getClusterAlias(), shard.getNodeId()),
+            request, context.getTask(), listener);
+    }
+
+    @Override
+    protected void onShardGroupFailure(int shardIndex, SearchShardTarget shardTarget, Exception exc) {
+        progressListener.notifyQueryFailure(shardIndex, shardTarget, exc);
+    }
+
+    @Override
+    protected void onShardResult(SearchPhaseResult result, SearchShardIterator shardIt) {
+        QuerySearchResult queryResult = result.queryResult();
+        if (queryResult.isNull() == false
+                // disable sort optims for scroll requests because they keep track of the last bottom doc locally (per shard)
+                && context.getRequest().scroll() == null
+                // top docs are already consumed if the query was cancelled or in error.
+                && queryResult.hasConsumedTopDocs() == false
+                && queryResult.topDocs() != null
+                && queryResult.topDocs().topDocs.getClass() == TopFieldDocs.class) {
+            TopFieldDocs topDocs = (TopFieldDocs) queryResult.topDocs().topDocs;
+            if (bottomSortCollector == null) {
+                synchronized (this) {
+                    if (bottomSortCollector == null) {
+                        bottomSortCollector = new BottomSortValuesCollector(topDocsSize, topDocs.fields);
+                    }
+                }
+            }
+            bottomSortCollector.consumeTopDocs(topDocs, queryResult.sortValueFormats());
+        }
+        super.onShardResult(result, shardIt);
+    }
+
+    @Override
+    protected SearchPhase getNextPhase(final SearchPhaseResults<SearchPhaseResult> results, SearchPhaseContext context) {
+        // TODO: refresh shard routing?
+        QueryPhaseResultConsumer resultConsumer = (QueryPhaseResultConsumer) results;
+        GroupShardsIterator<SearchShardIterator> shardsToRetry = getIterator(context);
+        if (shardsToRetry.size() > 0) {
+            return new RetryQueryPhase(context, context.getLogger(), searchPhaseController, resultConsumer, shardsToRetry);
```

Comment:
I'm wondering if we need a concrete class that takes care of re-executing the failed shard searches or we could just  reuse this class but provide only the retryable shards.